edb/lang/ir/ast.py

```diff
@@ -135,25 +135,43 @@ class EmptySet(Set):
     pass
 
 
-class Constant(Expr):
+class BaseConstant(Expr):
 
-    value: object
+    value: str
     type: s_types.Type
 
     def __init__(self, *args, type, **kwargs):
-        if type is None:
-            raise ValueError('type argument must not be None')
         super().__init__(*args, type=type, **kwargs)
+        if self.type is None:
+            raise ValueError('cannot create irast.Constant without a type')
+        if self.value is None:
+            raise ValueError('cannot create irast.Constant without a value')
 
 
-class StringConstant(Constant):
+class StringConstant(BaseConstant):
     pass
 
 
-class RawStringConstant(Constant):
+class RawStringConstant(BaseConstant):
     pass
 
 
+class IntegerConstant(BaseConstant):
+    value: str
+
+
+class FloatConstant(BaseConstant):
+    value: str
+
+
+class BooleanConstant(BaseConstant):
+    value: bool
```

Comment:
Sure you want `BooleanConstant` to be an exception and have a non-string `value`?

Reply:
Moreover, it seems that you always pass a string to it, not a bool.